File: Restful & Web Services/System Application/fire-extinguisher-mis/backend/auth-service/src/services/auth/auth.controller.ts
import type { Request, Response } from "express";
import { authService } from "./auth.service";
import { sendSuccess } from "../../shared/response";

export const authController = {
  async register(req: Request, res: Response) {
    const { user, tokens } = await authService.register(req.body);
    return sendSuccess(
      res,
      { user, ...tokens },
      "Registration successful",
      201
    );
  },

  async login(req: Request, res: Response) {
    const { user, tokens } = await authService.login(req.body);
    return sendSuccess(res, { user, ...tokens }, "Login successful");
  },

  async logout(req: Request, res: Response) {
    await authService.logout(req.body.refreshToken, req.user?.id);
    return sendSuccess(res, null, "Logout successful");
  },

  async refresh(req: Request, res: Response) {
    const tokens = await authService.refresh(req.body.refreshToken);
    return sendSuccess(res, tokens, "Token refreshed successfully");
  },

  async me(req: Request, res: Response) {
    const user = await authService.getMe(req.user!.id);
    return sendSuccess(res, user, "Current user retrieved");
  },

  async forgotPassword(req: Request, res: Response) {
    const result = await authService.forgotPassword(req.body.email);
    return sendSuccess(
      res,
      result,
      "If the email exists, a password reset token has been generated"
    );
  },

  async resetPassword(req: Request, res: Response) {
    await authService.resetPassword(req.body.token, req.body.newPassword);
    return sendSuccess(res, null, "Password has been reset successfully");
  },
};

File: Advanced Java & Spring Boot/System Application/utility-billing-system/src/main/java/com/ubs/controller/AuthController.java
package com.ubs.controller;

import com.ubs.dto.request.*;
import com.ubs.dto.response.AuthResponse;
import com.ubs.dto.response.SignupResponse;
import com.ubs.service.AuthService;
import io.swagger.v3.oas.annotations.Operation;
import io.swagger.v3.oas.annotations.tags.Tag;
import jakarta.validation.Valid;
import lombok.RequiredArgsConstructor;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;

import java.util.Map;

@RestController
@RequestMapping("/api/auth")
@RequiredArgsConstructor
@Tag(name = "Authentication")
public class AuthController {

    private final AuthService authService;

    @PostMapping("/signup")
    @Operation(summary = "Register a new user")
    public ResponseEntity<SignupResponse> signup(@Valid @RequestBody SignupRequest request) {
        return ResponseEntity.status(HttpStatus.CREATED).body(authService.signup(request));
    }

    @PostMapping("/login")
    @Operation(summary = "Login and receive JWT")
    public ResponseEntity<AuthResponse> login(@Valid @RequestBody LoginRequest request) {
        return ResponseEntity.ok(authService.login(request));
    }

    @PostMapping("/verify-otp")
    @Operation(summary = "Verify OTP for account or login")
    public ResponseEntity<Map<String, String>> verifyOtp(@Valid @RequestBody VerifyOtpRequest request) {
        authService.verifyOtp(request);
        return ResponseEntity.ok(Map.of("message", "OTP verified successfully"));
    }

    @PostMapping("/forgot-password")
    @Operation(summary = "Request password reset OTP")
    public ResponseEntity<Map<String, String>> forgotPassword(@Valid @RequestBody ForgotPasswordRequest request) {
        authService.forgotPassword(request);
        return ResponseEntity.ok(Map.of("message", "Password reset OTP sent to your email"));
    }

    @PostMapping("/reset-password")
    @Operation(summary = "Reset password with OTP")
    public ResponseEntity<Map<String, String>> resetPassword(@Valid @RequestBody ResetPasswordRequest request) {
        authService.resetPassword(request);
        return ResponseEntity.ok(Map.of("message", "Password reset successfully"));
    }
}
